xfrm: add rules for encrypt overlay

harsimran-pabla opened the pull request:
This pull requests updates XFRM policies to work with Encrypted Overlay.

When encryption.ipsec.EncryptedOverlay is enabled VXLAN traffic will be encrypted with IPSec/ESP. In this scenario our XFRM hooks will see VXLAN packets instead of pod-to-pod traffic.

Therefore, we must include XFRM policies and state which match on the source and destination of our VXLAN packets.
This will be the underlay addresses of each remote node.

When feature is disabled, rules added by this feature will be removed.


```release-note
ipsec: support EncryptedOverlay XFRM policies 
```

harsimran-pabla committed:
xfrm: provide req id in UpsertIPsecEndpoint

Provide req id parameter in UpsertIPsecEndpoint, which can be used to
override default req ID '1'. This is useful in cases where we install
feature specific xfrm rules and want graceful cleanup of those rules
when feature is disabled.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
workflow: update ipsec e2e tests to account for encrypt overlay rules

Encrypt overlay feature when enabled, will add 2 IPv4 xfrm state rules
per node. This change modifies the test script to account for additional
rules when this feature is enabled.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

danehans commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -213,24 +217,24 @@ func (n *linuxNodeHandler) enableIPsecIPv4(newNode *nodeTypes.Node, nodeID uint1
 				}
 
-				spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localCiliumInternalIP, remoteCiliumInternalIP, nodeID, newNode.BootID, ipsec.IPSecDirIn, zeroMark, updateExisting)
+				spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localCiliumInternalIP, remoteCiliumInternalIP, nodeID, newNode.BootID, ipsec.IPSecDirIn, zeroMark, updateExisting, ipsec.DefaultReqID)
```
Do you mean shadow `spi` and `err` inside the loop?

harsimran-pabla replied:
Both err and spi are used immediately for logging purpose. We do update `errs` if err is not nil. `upsertIPsecLog` returns err or logs debug message. I don't think there is shadowing concern, since `errs` will contain all returned errors. 

Usage of both SPI and err is not changed in this PR. Modification is adding additional parameter to `UpsertIPsecEndpoint`

pchaigno replied:
I'm probably the one who messed this up at some point :grimacing: No impact I can see, as Pabla said, but I'll try to send a fix anyway.

harsimran-pabla replied:
@pchaigno I can fix it in this PR as well, seems straight forward change.

harsimran-pabla replied:
Unless we want to backport that fix for some reason, then it is probably not a right PR to do it.

pchaigno replied:
I don't think we need to backport given the lack of impact, so feel free to make it in this PR. Thanks! :pray:

danehans commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -325,19 +359,19 @@ func (n *linuxNodeHandler) enableIPsecIPv6(newNode *nodeTypes.Node, nodeID uint1
 				}
 
-				spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localCiliumInternalIP, remoteCiliumInternalIP, nodeID, newNode.BootID, ipsec.IPSecDirIn, zeroMark, updateExisting)
+				spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localCiliumInternalIP, remoteCiliumInternalIP, nodeID, newNode.BootID, ipsec.IPSecDirIn, zeroMark, updateExisting, ipsec.DefaultReqID)
```
Do you mean shadow `spi` and `err` inside the loop?

danehans commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -349,13 +383,13 @@ func (n *linuxNodeHandler) enableIPsecIPv6(newNode *nodeTypes.Node, nodeID uint1
 				errs = errors.Join(errs, fmt.Errorf("failed to replace ipsec OUT (%q): %w", remoteCIDR.IP, err))
 			}
-			spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, remoteCIDR, localIP, remoteIP, nodeID, newNode.BootID, ipsec.IPSecDirOut, false, updateExisting)
+			spi, err := ipsec.UpsertIPsecEndpoint(wildcardCIDR, remoteCIDR, localIP, remoteIP, nodeID, newNode.BootID, ipsec.IPSecDirOut, false, updateExisting, ipsec.DefaultReqID)
```
Do you mean shadow `spi` and `err` inside the loop?

danehans commented on pkg/datapath/types/node.go:
```diff
@@ -75,6 +75,9 @@ type LocalNodeConfiguration struct {
 	EnableIPSec bool
 
+	// EnableIPSecEncryptedOverlay enables IPSec routes for overlay traffic
+	EnableIPSecEncryptedOverlay bool
```
Does this require EnableIPSec and/or tunnel mode to use VXLAN? If so, please update the godocs.

harsimran-pabla replied:
Added in the docs of the feature flag - latest commit.

danehans reviewed (approved):
A few non-blocking questions. Otherwise, /lgtm.

pchaigno commented on pkg/option/config.go:
```diff
@@ -2481,7 +2481,8 @@ func (c *DaemonConfig) TunnelingEnabled() bool {
 	return c.EnableNodePort || c.EnableHostFirewall || c.EnableWireguard ||
-		c.EnableHighScaleIPcache || c.EnableL2Announcements || c.ForceDeviceRequired
+		c.EnableHighScaleIPcache || c.EnableL2Announcements || c.ForceDeviceRequired ||
+		c.EnableIPSecEncryptedOverlay
```
Why are devices required for this feature? If IPsec is enabled, we will attach bpf_network to native devices. Isn't that enough?

harsimran-pabla replied:
From what I understood with discussion with @ldelossa, cil_to_netdev program is required for Encrypted Overlay feature to work. We are inspecting vxlan packet and deciding if it should be encrypted [here](https://github.com/cilium/cilium/blob/main/bpf/bpf_host.c#L1312-L1347). I don't think this BPF program is attached when just ipsec is enabled.

pchaigno commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -22,6 +22,10 @@ import (
 )
 
+var (
+	HostMask = net.IPv4Mask(255, 255, 255, 255)
```
It doesn't look like this is used outside the package. Also, how about making the name a bit more specific? Maybe `exactMatchMask`?

pchaigno commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -258,6 +262,36 @@ func (n *linuxNodeHandler) enableIPsecIPv4(newNode *nodeTypes.Node, nodeID uint1
+				remoteUnderlayIP := newNode.GetNodeIP(false)
+				if remoteUnderlayIP == nil {
+					return false, errs
+				}
```
Instead of retrieving this here a second time (and in a different way), how about we move `localNodeInternalIP`/`remoteNodeInternalIP` above (lines 205--210) outside of the if/else and use them in both the `subnetEncryption` case and this `encryptedOverlay` case?

harsimran-pabla replied:
I did more digging, `n.getV4LinkLocalIP()` does not return Node Internal IP. In case of tunnel configuration, it picks tunnel interface. 

https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec.go#L30-L55

pchaigno replied:
Isn't that a bug in `getLinkLocalIP`? cilium_vxlan doesn't have an IP address.

harsimran-pabla replied:
Yes, it seems so. I think this code path never gets executed in reality. We enter `subnetEncryption` when ipam mode is one of the cloud providers, in which case we are running native routing. In tunnel mode, we have other ipam modes ( cluster, k8s etc ) in which case we are using CiliumInternalIPs `! subnetEncryption`. 

I think bug fix is required - potentially in separate PR. I am thinking we might need some test coverage around it and backport it to previous releases. Once bug fix is in, we can optimize this code and move evaluation of `localNodeInternalIP/remoteNodeInternalIP` outside `subnetEncryption`.

harsimran-pabla replied:
Hi @pchaigno, any concerns in resolving this comment ?

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -67,6 +67,9 @@ const (
 	DefaultReqID = 1
+
+	// EncryptOverlayReqID is the reqid used for encrypting overlay traffic.
+	EncryptOverlayReqID = 2
```
Nit: `EncryptOverlayReqID` -> `EncryptedOverlayReqID`.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -930,6 +933,52 @@ func DeleteXfrm() error {
 }
 
+// DeleteXfrmWithReqID remove any XFRM policy or state from tables which matches the reqID
+func DeleteXfrmWithReqID(reqID int) error {
```
Nit: According to Go conventions, I believe this should be `DeleteXFRMWithReqID`.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -930,6 +933,52 @@ func DeleteXfrm() error {
+	}
+
+	return ee.Error()
+}
```
Instead of a new function, did you consider building this as a more general case of `DeleteXfrm` (which apparently also doesn't follow Go naming conventions :sweat_smile:)? We could have `DeleteXFRMWithReqID` with `reqID == 0` meaning we ignore the `reqID` and then `DeleteXFRM()` would simply return `DeleteXFRMWithReqID(0)`.

pchaigno commented on .github/actions/ipsec-key-rotate/action.yaml:
```diff
@@ -42,13 +46,23 @@ runs:
+          # If encryption-overlay is enabled, we have 4 more IPv4 keys for the overlay.
+          # Transition from single-key to per-tunnel keys goes like this 1 > 9(13) > 8(12).
+          # Transition from per-tunnel to single-key goes like this 8(12) > 9(13) > 1.
+          # Transition from per-tunnel keys to per-tunnel keys goes like this 8(12) > 16(24) > 8(12).
```
Thanks!

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -930,6 +933,52 @@ func DeleteXfrm() error {
 }
 
+// DeleteXfrmWithReqID remove any XFRM policy or state from tables which matches the reqID
+func DeleteXfrmWithReqID(reqID int) error {
```
To test this function, you could extend the `TestNodeChurnXFRMLeaks` integration test (with new configuration(s) to cover EncryptedOverlay). It currently exists to ensure that `DeleteXfrm` works as expected and that we don't leak XFRM states or policies.

harsimran-pabla replied:
No easy way to test Privileged tests locally, added new test case `TestEncryptedOverlayXFRMLeaks`. Will let CI run it for me.

pchaigno replied:
I usually use method 3 from https://docs.cilium.io/en/stable/contributing/testing/unit/#testing-individual-packages, but maybe you mean you're on macOS and it's different there?

harsimran-pabla replied:
Running them in linux vm (aarch64) on mac m2, I think there are some dependencies which are missing from my system. I wish we could create a containerised version of privilege tests. Start container with all dependencies and sandbox the test in it.

harsimran-pabla replied:
Actually it works 
`PRIVILEGED_TESTS=true go test -exec "sudo -E" ./pkg/datapath/linux -run=Test/linuxPrivilegedIPv4OnlyTestSuite/TestEncryptedOverlayXFRMLeaks`

It was genuine test failure.

ldelossa reviewed (approved):
Looks good to me.

pchaigno reviewed (commented):
Thanks for the neat PR! Very easy to review! :heart: 

One thing I don't understand (first comment), every other comments seem minor.

derailed reviewed (approved):
@harsimran-pabla Nice work! 
Just a note: This type of work would have been a bit easier and less invasive if some of the calls were to take in options vs 8+ args and possibly prevent some of the duplication...

ldelossa reviewed (changes requested):
Unfortunately this PR will not work as is. 

When testing this out, XFRM checks fail on the decrypted and recirculated VXLAN packet as it enters the stack for VXLAN decap. 

We need to block this PR until
1. We find the correct XFRM policy and state combination that allows the decrypted packet back into the stack for VXLAN decap
2. We define a custom output mark on encrypted overlay xfrm policies, look for this mark after decrypt, and redirect this packet in eBPF to a vxlan devices for decap.

ldelossa committed:
bpf: redirect decrypted overlay traffic

This commit updates bpf_host to identify decrypted overlay traffic and
redirect it to the VXLAN device for decapsulation.

After this decapsulation the original payload will be delivered to the
destination.

Without a redirect the decrypted packet comes back around into the stack
and the stack drops the packet in the XFRM hooks within the UDP receive
portion of the input path.

A redirect seems to clear the XFRM state associated with the skb and
allows the stack to process the packet as if the input device was the
VXLAN device.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

tommyp1ckles commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -553,14 +558,15 @@ func ipSecReplaceStateOut(localIP, remoteIP net.IP, nodeID uint16, localBootID,
 }
 
-func _ipSecReplacePolicyInFwd(src, dst *net.IPNet, tmplSrc, tmplDst net.IP, proxyMark bool, dir netlink.Dir) error {
+func _ipSecReplacePolicyInFwd(src, dst *net.IPNet, tmplSrc, tmplDst net.IP, proxyMark bool, dir netlink.Dir, reqID int) error {
```
nit/nonblocking/suggestion: while we're changing this, we might as well get rid of the C++ style underscore naming here, doesn't mean much in Go.

harsimran-pabla replied:
There are a couple of styling improvements required - removing underscore and passing a struct as a parameter since param list is getting long. I was thinking of doing this as an independent PR, so code style can be the focus of that review.

ldelossa replied:
Agree, independent PR is good, as some of the other style changes in the PR are being pushed back against for creating noise.

tommyp1ckles commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -214,24 +214,24 @@ func (n *linuxNodeHandler) enableIPsecIPv4(newNode *nodeTypes.Node, nodeID uint1
 
 			for _, cidr := range n.nodeConfig.IPv4PodSubnets {
-				spi, err = ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localIP, remoteIP, nodeID, newNode.BootID, ipsec.IPSecDirOut, zeroMark, updateExisting)
+				spi, err = ipsec.UpsertIPsecEndpoint(wildcardCIDR, cidr, localIP, remoteIP, nodeID, newNode.BootID, ipsec.IPSecDirOut, zeroMark, updateExisting, ipsec.DefaultReqID)
```
I'm a little bit confused as to how these ipsec changes are related to the commits purpose of skipping un-needed endpoint logs.

ldelossa replied:
@tommyp1ckles So you understand the need for this as a whole, to remove only EncryptedOverlay XFRM policies when this feature is disabled, and IPSec feature as hole remains? And rather, you think the commit in which the change to plumb the ReqID (the means to do the former) is the wrong place for it?

harsimran-pabla replied:
Hmm, [this commit](https://github.com/cilium/cilium/pull/31757/commits/cb14fdba63432da61cdcd42ce80ba4cdd63b60b6) looks bit out of place. @ldelossa  is it possible something got squashed incorrectly and ipsec req id change got fused with endpoint log commit ?

ldelossa replied:
Ah! I didn't even realize that. Yes this looks odd. I imagine we can drop this as long as the other reqID diff gets in there appropriately. 

Yeah. probably rebase nonsense.

ldelossa replied:
Okay that snafu should be fixed, sorry about that!!

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1058,6 +1058,21 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
+		 *
+		 * If we see this here, its decrypted overlay traffic from these
+		 * XFRM policies. We can punt this directly to ingress side of vxlan
+		 * interface for decrypt.
```
```suggestion
		 * If we see this here, its decrypted overlay traffic from these
		 * XFRM policies. We can punt this directly to ingress side of vxlan
		 * interface for decap.
```

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1058,6 +1058,21 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
 #ifdef ENABLE_IPSEC
+#ifdef ENABLE_ENCRYPTED_OVERLAY
+	if (!from_host) {
+		/* EncryptedOverlay XFRM policies set the output mark to the literal
```
Can you please fold this into the other `!from_host` case below?

julianwiedmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -395,7 +395,7 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.EnableIPsecKeyWatcher)
 
-	flags.Bool(option.EnableIPSecEncryptedOverlay, defaults.EnableIPSecEncryptedOverlay, "Enable IPsec encrypted overlay. If enabled tunnel traffic will be encrypted before leaving the host.")
+	flags.Bool(option.EnableIPSecEncryptedOverlay, defaults.EnableIPSecEncryptedOverlay, "Enable IPSec encrypted overlay. If enabled tunnel traffic will be encrypted before leaving the host. Requires ipsec and tunnel mode vxlan to be enabled.")
```
The `s/IPsec/IPSec` change feels very arbitrary, and isn't correct (the feature really is called `IPsec`, sorry).

ldelossa replied:
Sure, I'll back this out.

julianwiedmann commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -31,6 +31,14 @@ const (
 	// needs to decrypt a packet.
 	RouteMarkDecrypt = 0x0D00
 
+	// RouteMarkDecryptedOverlay is the output mark used for EncryptedOverlay
+	// XFRM policies.
+	//
+	// When this mark is present on a packet it indicates that overlay traffic
+	// was decrypted by XFRM and should be forwarded to a tunnel device for
+	// decapsulation.
+	RouteMarkDecryptedOverlay = 0x1D00
+
```
This belongs into `pkg/datapath/linux/linux_defaults/mark.go`, right?

ldelossa replied:
I'm not sure, all the other XFRM/IPSec related marks are also defined in this file:

```
	// RouteMarkDecrypt is the default route mark to use to indicate datapath
	// needs to decrypt a packet.
	RouteMarkDecrypt = 0x0D00

	// RouteMarkDecryptedOverlay is the output mark used for EncryptedOverlay
	// XFRM policies.
	//
	// When this mark is present on a packet it indicates that overlay traffic
	// was decrypted by XFRM and should be forwarded to a tunnel device for
	// decapsulation.
	RouteMarkDecryptedOverlay = 0x1D00

	// RouteMarkEncrypt is the default route mark to use to indicate datapath
	// needs to encrypt a packet.
	RouteMarkEncrypt = 0x0E00
```

julianwiedmann replied:
I would assume that's because those marks are then also used in routing rules. But fair enough.

julianwiedmann commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -177,7 +177,7 @@ func setupLinuxPrivilegedIPv4AndIPv6TestSuite(tb testing.TB) *linuxPrivilegedIPv
 }
 
 func tearDownTest(tb testing.TB) {
-	ipsec.DeleteXfrm()
+	ipsec.DeleteXFRM()
```
Are these changes really needed? Imho it's just creating unrelated noise.

ldelossa replied:
This was requested by Paul, dont want to get into the back and forth about naming, its in the commit now so we can probably just leave it, no?

julianwiedmann replied:
Sure. The usual way of handling such cosmetic changes is to make your reviewers' life easy, and split them off into a separate patch :).

ldelossa replied:
Going to resolve for now, since this was a request. If you find it a larger issue feel free to ping me and I'll re-open and revert.

ldelossa replied:
> Sure. The usual way of handling such cosmetic changes is to make your reviewers' life easy, and split them off into a separate patch :).

Yup I agree, I think we'll push back on cosmetic changes in future PRs and push to separate patch.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1058,6 +1058,21 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
+		 * XFRM policies. We can punt this directly to ingress side of vxlan
+		 * interface for decrypt.
+		 */
+		if (ctx->mark == MARK_MAGIC_DECRYPTED_OVERLAY) {
+			ret = ctx_redirect(ctx, ENCAP_IFINDEX, BPF_F_INGRESS);
+			return ret;
+		}
```
Could you motivate why the manual redirect is needed? What makes this overlay traffic different from "normal" inbound VXLAN traffic - which would just run until the end of `from-netdev` and get passed to the stack with `CTX_ACT_OK`?

ldelossa replied:
Without a redirect the decrypted packet comes back around into the stack and the stack drops the packet in the XFRM hooks within the UDP receive portion of the input path. 

A redirect seems to clear the XFRM state associated with the skb and allows the stack to process the packet as if the input device was the VXLAN device. 

I and Gray took a stab at fixing this but didn't come to a conclusion for getting the decryped packet past UDP. Its possible another solution exists since Pabla discovered that avoiding the zero'ing out of a the post decryped packet marks seems to fix the issue, but this would require a larger change to current IPSec flow and I don't want to do this in the initial implementation of EncryptedOverlay. Given a bit more time and testing, I want to explore that separately.

julianwiedmann replied:
You'll appreciate that all this context belongs into either a comment, or the patch description itself. So that the next person doesn't start from scratch :).

It's rather annoying that we now have **three** different ways of handling overlay traffic in `from-netdev`:
1. pass it up the stack
2. redirect, but [strip the VXLAN headers](https://github.com/cilium/cilium/blob/d18d41e6995679624980671c46b25e5a80d40136/bpf/bpf_host.c#L1285)
3. redirect, without stripping the VXLAN headers (new)

ldelossa replied:
I will add the follow to the commit message:

>Without a redirect the decrypted packet comes back around into the stack and the stack drops the packet in the XFRM hooks within the UDP receive portion of the input path.

>A redirect seems to clear the XFRM state associated with the skb and allows the stack to process the packet as if the input device was the VXLAN device.

ldelossa replied:
> It's rather annoying that we now have three different ways of handling overlay traffic in from-netdev:

Yeah I agree. 

We want to support Encrypted Overlay for all unicast traffic. I want to push that forward in the upcoming weeks. When this effort begins I can address this change once again and see if we can limit these cases.

julianwiedmann reviewed (changes requested):
A few smaller changes / asks for clarification. I didn't look at the XFRM changes in detail.

julianwiedmann commented on pkg/datapath/linux/node.go:
```diff
@@ -1294,11 +1294,17 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
 			log.WithError(err).Warning("Cannot cleanup previous encryption rule state.")
 		}
-		if err := ipsec.DeleteXfrm(); err != nil {
+		if err := ipsec.DeleteXFRM(); err != nil {
 			return fmt.Errorf("failed to delete xfrm policies on node configuration changed: %w", err)
 		}
 	}
 
+	if !newConfig.EnableIPSecEncryptedOverlay {
+		if err := ipsec.DeleteXFRMWithReqID(ipsec.EncryptedOverlayReqID); err != nil {
+			return fmt.Errorf("failed to delete encrypt overlay xfrm policies on node configuration change: %w", err)
+		}
+	}
+
```
Please take a follow-on item to see if you can optimize this XFRM processing, being mindful of https://github.com/cilium/cilium/pull/32588.

ldelossa replied:
Will do.

julianwiedmann reviewed (approved):
I looked at the datapath integration, and that's cool. Please see one follow-on issue noted below, and I'm looking forward to unified VXLAN handling in `from-netdev` ;).

Didn't look at the XFRM integration in detail, or whether this is the best-possible design.

pchaigno commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -826,6 +830,75 @@ func TestNodeChurnXFRMLeaks(t *testing.T) {
+	policies, err = netlink.XfrmPolicyList(netlink.FAMILY_ALL)
+	require.NoError(t, err)
+	require.Equal(t, 1, countXFRMPolicies(policies))
+}
```
This was added in the first commit by mistake, right? Or I'm not seeing the link between this test and the commit description :confused:
